File: 240703/정수 명령 처리 8/process-numeric-commands-8.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.node_num = 0

    def push_front(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head

        if self.head != None:
            self.head.prev = new_node
            self.head = new_node
            new_node.prev = None
        
        else:
            self.head = new_node
            self.tail = new_node
            new_node.prev = None

        self.node_num += 1
    
    def push_back(self, new_data):
        new_node = Node(new_data)
        new_node.prev = self.tail

        if self.tail != None:
            self.tail.next = new_node
            self.tail = new_node
            new_node.next = None
        
        else:
            self.head = new_node
            self.tail = new_node
            new_node.next = None
        
        self.node_num += 1

    def pop_front(self):

        if self.head == None:
            print("List is Empty")
        
        elif self.head.next == None:
            temp = self.head

            self.head = None
            self.tail = None
            self.node_num = 0
            return temp.data
        
        else:
            temp = self.head
            temp.next.prev = None
            self.head = temp.next
            temp.next = None

            self.node_num -= 1
            return temp.data
        
    def pop_back(self):
        
        if self.tail == None:
            print("List is Empty")
        
        elif self.tail.prev == None:
            temp = self.tail

            self.head = None
            self.tail = None
            self.node_num = 0

            return temp.data
        
        else:
            temp = self.tail
            temp.prev.next = None
            self.tail = temp.prev
            temp.prev = None

            self.node_num -= 1
            return temp.data
        
    def size(self):
        return self.node_num
    
    def empty(self):
        return self.node_num == 0

    def front(self):
        if self.head == None:
            print("List is empty")
        else:
            return self.head.data
    
    def back(self):
        if self.tail == None:
            print("List is empty")
        else:
            return self.tail.data


n = int(input())

l = DoublyLinkedList()

for _ in range(n):
    line = list(input().split())

    if line[0] == "push_front":
        l.push_front(int(line[1]))
    
    elif line[0] == "push_back":
        l.push_back(int(line[1]))

    elif line[0] == "pop_front":
        print(l.pop_front())
    
    elif line[0] == "pop_back":
        print(l.pop_back())

    elif line[0] == "size":
        print(l.size())

    elif line[0] == "empty":
        if l.empty():
            print(1)
        else:
            print(0)
    
    elif line[0] == "front":
        print(l.front())
    
    elif line[0] == "back":
        print(l.back())

File: 240907/정수 명령 처리 8/process-numeric-commands-8.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct node{
    int data;
    struct node *prev;
    struct node *next; 
} node;

typedef struct doubly_list
{
    node *head;
    node *tail;
    int size;
} doubly_list;

// 구조체의 포인터에 접근할 때는 ->를 사용
void    init(doubly_list *list)
{
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

void    push_back(doubly_list *list, int data)
{
    // 새로운 Node를 생성, 주소로 받음
    node *new_node = (node *)malloc(sizeof (node));
    
    new_node->data = data;
    new_node->next = NULL;

    // Doubly_list에 Node가 있는지 판단
    if (list->size == 0)
    {
        list->tail = new_node;
        list->head = new_node;
        new_node->prev = NULL;
        new_node->next = NULL;
        list->size++;
    }
    else
    {
        new_node->prev = list->tail;
        list->tail->next = new_node;
        list->tail = new_node;
        list->size++;
    }
}

void    push_front(doubly_list *list, int data)
{
    node *new_node = (node *)malloc(sizeof(node));

    new_node->data = data;
    new_node->prev = NULL;

    if (list->size == 0)
    {
        new_node->next = NULL;
        list->head = new_node;
        list->tail = new_node;
        list->size++;
    }
    else
    {
        new_node->next = list->head;
        list->head->prev = new_node;
        list->head = new_node;
        list->size++;
    }
}

void    pop_front(doubly_list *list)
{
    if (list->size == 0)
        return ;
    else if (list->size == 1)
    {
        printf("%d\n", list->head->data);
        list->head = NULL;
        list->tail = NULL;
        list->size--;
    }
    else {
        printf("%d\n", list->head->data);
        list->head = list->head->next;
        list->head->prev = NULL;
        list->size--;
    }
}

void    pop_back(doubly_list *list)
{
    if (list->size == 0)
        return ;
    else if (list->size == 1)
    {
        printf("%d\n", list->tail->data);
        list->head = NULL;
        list->tail = NULL;
        list->size--;
    }
    else
    {
        printf("%d\n", list->tail->data);
        list->tail = list->tail->prev;
        // list->tail->next->prev = NULL;
        list->tail->next = NULL;
        list->size--;
    }
    return ;
}

void    size(doubly_list *list)
{
    printf("%d\n", list->size);
    return ;
}

void    empty(doubly_list *list)
{
    if (list->size)
        printf("0\n");
    else
        printf("1\n");
    return ;
}

void    front(doubly_list *list)
{
    if (list->size)
        printf("%d\n", list->head->data);
    else
        return ;
}

void    back(doubly_list *list)
{
    if (list->size)
        printf("%d\n", list->tail->data);
    else
        return ;
}

int main() {
    
    int n;
    char s[20];
    int d;
    doubly_list d_list;

    scanf("%d", &n);
    init(&d_list);
    for (int i = 0; i < n; i++)
    {
        scanf("%s", s);
        if (strcmp(s, "push_front") == 0)
        {   
            scanf("%d", &d);
            push_front(&d_list, d);
        }
        else if (strcmp(s, "push_back") == 0)
        {
            scanf("%d", &d);
            push_back(&d_list, d);
        }
        else if (strcmp(s, "pop_front") == 0)
            pop_front(&d_list);
        else if (strcmp(s, "pop_back") == 0)
            pop_back(&d_list);
        else if (strcmp(s, "size") == 0)
            size(&d_list);
        else if (strcmp(s, "empty") == 0)
            empty(&d_list);
        else if (strcmp(s, "front") == 0)
            front(&d_list);
        else if (strcmp(s, "back") == 0)
            back(&d_list);
    }
    return 0;
}